server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OnHeapSingleTreeBuilder.java

```diff
@@ -0,0 +1,204 @@
+        for (int i = 0; i < numDocs; i++) {
+            starTreeDocuments[i] = getNextSegmentStarTreeDocument();
+        }
+        return processStarTreeDocuments(starTreeDocuments);
+    }
+
+    /**
+     * Sort, aggregates and merges the star-tree documents
+     * @param starTreeDocuments star-tree documents
+     * @return iterator for star-tree documents
+     * @throws IOException throws when unable to sort, merge and aggregate star-tree documents
+     */
+    public Iterator<StarTreeDocument> processStarTreeDocuments(StarTreeDocument[] starTreeDocuments) throws IOException {
+
+        // sort the documents
+        Arrays.sort(starTreeDocuments, (o1, o2) -> {
+            for (int i = 0; i < numDimensions; i++) {
+                if (o1.dimensions[i] != o2.dimensions[i]) {
+                    return Math.toIntExact(o1.dimensions[i] - o2.dimensions[i]);
+                }
+            }
+            return 0;
+        });
+
+        // merge the documents
+        return mergeStarTreeDocuments(starTreeDocuments);
+    }
+
+    /**
+     * Merges the star-tree documents
+     * @param starTreeDocuments star-tree documents
+     * @return iterator to aggregate star-tree documents
+     */
+    private Iterator<StarTreeDocument> mergeStarTreeDocuments(StarTreeDocument[] starTreeDocuments) {
+        return new Iterator<>() {
+            boolean hasNext = true;
+            StarTreeDocument currentStarTreeDocument = starTreeDocuments[0];
+            int docId = 1;
+
+            @Override
+            public boolean hasNext() {
+                return hasNext;
+            }
+
+            @Override
+            public StarTreeDocument next() {
+                // aggregate as we move on to the next doc
+                StarTreeDocument next = aggregateStarTreeDocument(null, currentStarTreeDocument);
+                while (docId < starTreeDocuments.length) {
+                    StarTreeDocument starTreeDocument = starTreeDocuments[docId++];
+                    if (!Arrays.equals(starTreeDocument.dimensions, next.dimensions)) {
+                        currentStarTreeDocument = starTreeDocument;
+                        return next;
+                    } else {
+                        next = aggregateStarTreeDocument(next, starTreeDocument);
+                    }
+                }
+                hasNext = false;
+                return next;
+            }
+        };
+    }
+
+    /**
+     * Generates a star-tree for a given star-node
+     * @param startDocId  Start document id in the star-tree
+     * @param endDocId    End document id (exclusive) in the star-tree
+     * @param dimensionId Dimension id of the star-node
+     * @return iterator for star-tree documents of star-node
+     * @throws IOException throws when unable to generate star-tree for star-node
+     */
+    @Override
+    public Iterator<StarTreeDocument> generateStarTreeForStarNode(int startDocId, int endDocId, int dimensionId) throws IOException {
```

Comment:
nit : getStarTreeDocumentsForStarNode ?

is there star tree getting created for star node - if not lets fix the comment